File: offline-package/start.js
#!/usr/bin/env node
// IS630 Study Guide - offline launcher (Node fallback if Python is unavailable).
// Serves ./site on a local port and opens the browser. No internet needed.
const http = require('http')
const fs = require('fs')
const path = require('path')
const { exec } = require('child_process')

const PORT = 8630
const SITE = path.join(__dirname, 'site')

const MIME = {
  '.html': 'text/html', '.js': 'text/javascript', '.mjs': 'text/javascript',
  '.css': 'text/css', '.json': 'application/json', '.wasm': 'application/wasm',
  '.whl': 'application/octet-stream', '.zip': 'application/zip',
  '.svg': 'image/svg+xml', '.png': 'image/png', '.woff2': 'font/woff2',
}

if (!fs.existsSync(SITE)) {
  console.error("ERROR: 'site' folder not found next to start.js")
  process.exit(1)
}

http.createServer((req, res) => {
  let urlPath = decodeURIComponent((req.url || '/').split('?')[0])
  if (urlPath === '/') urlPath = '/index.html'
  const filePath = path.join(SITE, urlPath)
  // basic path-traversal guard
  if (!filePath.startsWith(SITE)) { res.writeHead(403); res.end(); return }
  fs.readFile(filePath, (err, data) => {
    if (err) { res.writeHead(404); res.end('Not found'); return }
    res.writeHead(200, {
      'Content-Type': MIME[path.extname(filePath)] || 'application/octet-stream',
      'Cache-Control': 'no-store',
    })
    res.end(data)
  })
}).listen(PORT, '127.0.0.1', () => {
  const url = `http://localhost:${PORT}/`
  console.log('='.repeat(56))
  console.log('  IS630 Study Guide is running offline')
  console.log(`  Open:  ${url}`)
  console.log('  (Press Ctrl+C here to stop the server)')
  console.log('='.repeat(56))
  const cmd = process.platform === 'win32' ? `start ${url}`
    : process.platform === 'darwin' ? `open ${url}` : `xdg-open ${url}`
  exec(cmd, () => {})
})

File: offline-package/start.py
#!/usr/bin/env python3
"""IS630 Study Guide - offline launcher.

Serves the bundled ./site folder on a local port and opens your browser.
Works on Windows / macOS / Linux with any Python 3. No internet needed.

Usage:  python start.py   (or double-click start-mac.command / start-windows.bat)
"""
import http.server
import socketserver
import webbrowser
import os
import sys
from functools import partial

PORT = 8630
SITE = os.path.join(os.path.dirname(os.path.abspath(__file__)), "site")

if not os.path.isdir(SITE):
    print("ERROR: 'site' folder not found next to start.py")
    sys.exit(1)


class Handler(http.server.SimpleHTTPRequestHandler):
    # Correct MIME types so the browser accepts the WASM runtime and wheels.
    extensions_map = {
        **http.server.SimpleHTTPRequestHandler.extensions_map,
        ".wasm": "application/wasm",
        ".mjs": "text/javascript",
        ".js": "text/javascript",
        ".json": "application/json",
        ".whl": "application/octet-stream",
        ".zip": "application/zip",
    }

    def end_headers(self):
        # Never cache during a session, and keep it snappy.
        self.send_header("Cache-Control", "no-store")
        super().end_headers()

    def log_message(self, *args):
        pass  # quiet


def main():
    handler = partial(Handler, directory=SITE)
    socketserver.TCPServer.allow_reuse_address = True
    with socketserver.TCPServer(("127.0.0.1", PORT), handler) as httpd:
        url = f"http://localhost:{PORT}/"
        print("=" * 56)
        print("  IS630 Study Guide is running offline")
        print(f"  Open:  {url}")
        print("  (Press Ctrl+C here to stop the server)")
        print("=" * 56)
        try:
            webbrowser.open(url)
        except Exception:
            pass
        try:
            httpd.serve_forever()
        except KeyboardInterrupt:
            print("\nStopped.")


if __name__ == "__main__":
    main()
